data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
+  - Module `List` through `List.iter` and the type of its `f` function
+  - Module `Out_channel` through `Out_channel.output_string`
+
+Check the behaviour of the program using `dune exec funkt < dune`.
+
+**Dependency Injection**
+
+[Dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is a way to parametrise over a dependency.
+
+Here is a refactoring of the module `IterPrint` to make of this technique:
+
+**`iterPrint.ml`**
+```ocaml
+module type Iterable = sig
+  type 'a t
+  val iter : ('a -> unit) -> 'a t -> unit
+end
+
+module type S = sig
+  type 'a t
+  val f : string t -> unit
+end
+
+module Make(Dep: Iterable) : S with type 'a t := 'a Dep.t = struct
+  let f = Dep.iter (fun s -> Out_channel.output_string stdout (s ^ "\n"))
+end
+```
+
+The module `IterPrint` is refactored into a functor that takes the dependency providing `iter` as a parameter. The `with type 'a t := 'a Dep.t` constraint means the type `t` from the parameter `Dep` replaces the type `t` in the result module. This allows the type of `f` to use the type `t` from the parameter module `Dep`. With this refactoring, `IterPrint` only has one dependency; at the time it is compiled, no implementation of function `iter` is available yet.
+
+**Note**: An OCaml interface file must be a module, not a functor. Functors must be embedded inside modules. Therefore, it is customary to call them `Make`.
+
+**`funkt.ml`**
+
+```ocaml
+module StringSet = Set.Make(String)
+module IterPrint = IterPrint.Make(List)
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.elements
+  |> IterPrint.f
+```
+
+The dependency `List` is _injected_ when compiling the module `Funkt`. Observe that the code using `IterPrint` is unchanged. Check the program's behaviour using `dune exec funkt < dune`.
+
+**Replacing a Dependency**
+
+Now, replacing the implementation of `iter` inside `IterListPrint` is no longer a refactoring; it is another functor application with another dependency. Here, `Array` replaces `List`:
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+module IterPrint = IterPrint.Make(Array)
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.elements
+  |> Array.of_list
+  |> IterPrint.f
+```
+
+Check the program's behaviour using `dune exec funkt < dune`.
+
+**Note**: The functor `IterPrint.Make` returns a module that exposes the type of the injected dependency (here first `List.t` then `Array.t`). That's why a `with type` constraint is needed. If the dependency was an _implementation detail_ that is not exposed in the signature of the initial version of `IterMake` (i.e., in the type of `IterMake.f`), that constraint wouldn't be needed, and the call site of `IterPrint.f` would be unchanged when injecting another dependency.
```

Comment:
```suggestion
**Note**: The functor `IterPrint.Make` returns a module that exposes the injected dependency type (here, first `List.t` then `Array.t`). That's why a `with type` constraint is needed. If the dependency was an _implementation detail_ that wasn't exposed in the signature of `IterMake`'s initial version (i.e., in the type of `IterMake.f`), that constraint wouldn't be needed, and the call site of `IterPrint.f` would be unchanged when injecting another dependency.
```

If this isn't accurate, let's work on the syntax together!

Reply:
Check updated text line 354

Reply:
This is a long sentence with a lot of prepositional phrases: "If the dependency was an _implementation detail_ that is not exposed in the signature of the initial version of `IterMake` (i.e., in the type of `IterMake.f`), that constraint wouldn't be needed, and the call site of `IterPrint.f` would be unchanged when injecting another dependency."

I tried to remove one of them: 
"in the signature of the initial version of `IterMake`"
to
"in the  signature of `IterMake`'s initial version"

It doesn't shorted it much, but it tightens that a little bit. It is correct as it is, but this suggestion makes it a little tighter. Feel free to disregard. 🙂